File: komnaty/komnaty.py
def get_row(grid, n: int) -> list:
    """Returns a row"""
    return grid[n - 1]


def get_col(grid, n: int) -> list:
    """Returns a column"""
    return [row[n - 1] for row in grid]


def get_cell(grid, y: int, x: int) -> str:
    """Returns the value of a cell"""
    return grid[y - 1][x - 1]


def get_adjacent_x(grid, y: int, x: int) -> list:
    row = get_row(grid, y)
    positions = []

    for _x, cell in enumerate(row[x - 1:], start=x):
        current = (y, _x)
        if cell == "#":
            break

        positions.append(current)

    for _x, cell in zip(range(x, 0, -1), reversed(row[:x])):
        current = (y, _x)
        if cell == "#":
            break
            
        positions.append(current)
    return list(set(positions))


def get_adjacent_y(grid, y: int, x: int) -> list:
    col = get_col(grid, x)
    positions = []

    for _y, cell in enumerate(col[y - 1:], start=y):
        current = (_y, x)
        if cell == "#":
            break

        positions.append(current)

    for _y, cell in zip(range(y, 0, -1), reversed(col[:y])):
        current = (_y, x)
        if cell == "#":
            break
        
        positions.append(current)
    return list(set(positions))


def get_adjancent_cells(grid, y: int, x: int) -> list:
    # Getting the adjacent cells in the x axis
    adjacent_x = get_adjacent_x(grid, y, x)
    # Getting the adjacent cells in the y axis
    adjacent_y = get_adjacent_y(grid, y, x)
    return list(set([*adjacent_x, *adjacent_y]))


def get_surrounding_positions(grid, y: int, x: int) -> list:
    # Getting the positions
    positions = [(y + 1, x), (y, x - 1), (y, x + 1), (y - 1, x)]
    # Returning only valid positions
    return [(y, x) for (y, x) in positions if (y > 0 and x > 0) and (y <= len(grid) and x <= len(grid))]


def get_chamber(grid, y: int, x: int) -> list:
    # Getting the initial positions
    positions = list(set([(y, x), *get_adjancent_cells(grid, y, x)]))

    for pos in positions:
        surroundings = get_adjancent_cells(grid, *pos)

        for s in surroundings:
            if s not in positions:
                positions.append(s)

    return list(set(positions))


def get_chamber_count(grid, chamber: list) -> int:
    count = 0

    for room in chamber:
        surroundings = get_surrounding_positions(grid, *room)

        for cell in surroundings:
            value = get_cell(grid, *cell)

            if value == "#":
                count += 1
    
    return count


def main():
    grid, free_spaces = [], []
    n = int(input("> "))

    for i in range(n):
        row = list(input())
        grid.append(row)

    for i in range(n):
        y, x = input("> ").split()
        free_spaces.append((int(y), int(x)))

    for space in free_spaces:
        chamber = get_chamber(grid, *space)
        count = get_chamber_count(grid, chamber)
        print(count)


if __name__ == "__main__":
    main()

File: komnaty/komnaty.js
const prompt = require("prompt-sync")();

const grid = [
    ["#", "#", "#", "#", "#"],
    ["#", ".", ".", "#", "#"],
    ["#", ".", "#", ".", "#"],
    ["#", "#", "#", ".", "#"],
    ["#", "#", "#", "#", "#"]    
]

const free_spaces = [
    [2, 2], [2, 3], [3, 4], [4, 4], [2, 2]
]

function delete_duplicates(arr) {
    let uniques = [];
    let items_found = {};

    for (value of arr) {
        let stringified = JSON.stringify(value);
        if (items_found[value]) {
            continue;
        }

        uniques.push(value);
        items_found[value] = true;
    }
    return uniques;
}

function get_row(grid, n) {
    return grid[n - 1];
}

function get_col(grid, n) {
    let col = [];

    for (row of grid) {
        col.push(row[n - 1]);
    }
    return col;
}

function get_cell(grid, y, x) {
    return grid[y - 1][x - 1];
}

function get_adjacent_x(grid, y, x) {
    let row = get_row(grid, y);
    let positions = [];

    for (let i = x; i < row.length; i++) {
        let current = [y, i];
        let cell = row[i - 1];

        if (cell == "#") {
            break;
        }

        positions.push(current);
    }
    // Reversing
    for (let i = x; i > 0; i--) {
        let current = [y, i];
        let cell = row[i - 1];

        if (cell == "#") {
            break;
        }

        positions.push(current);
    }
    
    return delete_duplicates(positions);
}

function get_adjacent_y(grid, y, x) {
    let col = get_col(grid, x);
    let positions = [];

    for (let i = y; i < col.length; i++) {
        let current = [i, x];
        let cell = col[i - 1];

        if (cell == "#") {
            break;
        }

        positions.push(current);
    }
    // Reversing
    for (let i = y; i > 0; i--) {
        let current = [i, x];
        let cell = col[i - 1];

        if (cell == "#") {
            break;
        }

        positions.push(current);
    }
    
    return delete_duplicates(positions);
}

function get_adjacent_cells(grid, y, x) {
    let adjacent_x = get_adjacent_x(grid, y, x);
    let adjacent_y = get_adjacent_y(grid, y, x);
    adjacent_x.push(...adjacent_y);

    return delete_duplicates(adjacent_x);
}

function get_chamber(grid, y, x) {
//    const is_not_in_grid = (grid, y, x) => {
//        for ([_y, _x] of grid) {
//            if (_y === y && _x === x) {
//                return true;
//            }
//        }
//        return false;
//    }
//
//    let positions = get_adjacent_cells(grid, y, x);
//    positions.push([y, x]);
//
//    for (pos of positions) {
//        let surroundings = get_adjacent_cells(grid, ...pos);
//
//        for (s of surroundings) {
//            if (is_not_in_grid(grid, ...s)) {
//                positions.push(get_adjacent_cells(grid, ...s));
//            }
//        }
//    }
//    return delete_duplicates(positions);
}

console.log(get_chamber(grid, 3, 2));
